database/ticket_test.go

```diff
@@ -332,3 +362,110 @@ func testCountTickets(t *testing.T) {
+		t.Fatal("expected error for no ticket in db")
+	}
+
+	// Insert a ticket into the database.
+	err := db.InsertNewTicket(ticket)
+	if err != nil {
+		t.Fatalf("error storing ticket in database: %v", err)
+	}
+
+	// First three inserts are fine.
+	if err := db.AddAltSigHistory(ticket.Hash, tAltHist()); err != nil {
+		t.Fatalf("unexpected error for first alt history: %v", err)
+	}
+
+	// Same history is also fine.
+	sameHist := tAltHist()
+	if err := db.AddAltSigHistory(ticket.Hash, sameHist); err != nil {
+		t.Fatalf("unexpected error for second alt history: %v", err)
+	}
+	if err := db.AddAltSigHistory(ticket.Hash, sameHist); err != nil {
+		t.Fatalf("unexpected error for third alt history: %v", err)
+	}
+
+	// Over the max, currently 3, should error.
+	if err := db.AddAltSigHistory(ticket.Hash, sameHist); err == nil {
+		t.Fatalf("expected error for fourth alt history: %v", err)
+	}
+	if err := db.AddAltSigHistory(ticket.Hash, tAltHist()); err == nil {
+		t.Fatalf("expected error for second fourth alt history: %v", err)
+	}
+}
+
+func testAltSigHistory(t *testing.T) {
+	ticket := exampleTicket()
+
+	// No ticket means no history history.
+	history, err := db.AltSigHistory(ticket.Hash)
+	if err != nil {
+		t.Fatalf("unexpected error for no ticket in db: %v", err)
+	}
+	if len(history) != 0 {
+		t.Fatal("expected no history for no ticket in db")
+	}
+
+	// Insert a ticket into the database.
+	err = db.InsertNewTicket(ticket)
+	if err != nil {
+		t.Fatalf("error storing ticket in database: %v", err)
+	}
+
+	// No history added yet.
+	history, err = db.AltSigHistory(ticket.Hash)
+	if err != nil {
+		t.Fatalf("unexpected error for new: %v", err)
+	}
+	if len(history) != 0 {
+		t.Fatal("expected no history for new ticket")
+	}
+
+	// First three additions should return the correct appended history.
+	wantHist := make([]*AltSigHistory, 0, 3)
+	for i := 0; i < 3; i++ {
+		altHist := tAltHist()
+		wantHist = append(wantHist, altHist)
+		if err := db.AddAltSigHistory(ticket.Hash, altHist); err != nil {
+			t.Fatalf("unexpected error for first alt history: %v", err)
+		}
+
+		history, err = db.AltSigHistory(ticket.Hash)
+		if err != nil {
+			t.Fatalf("unexpected error for alt history %d: %v", i+1, err)
+		}
+		if len(history) != i+1 {
+			t.Fatalf("expected history len of %d but got %d", i+1, len(history))
+		}
+		if !reflect.DeepEqual(wantHist, history) {
+			t.Fatalf("expected history %v but got %v", spew.Sdump(wantHist), spew.Sdump(history))
+		}
+	}
+
+	// Further additions should error and not change the history.
+	if err := db.AddAltSigHistory(ticket.Hash, tAltHist()); err == nil {
+		t.Fatalf("expected error for forth alt history addition")
+	}
+
+	history, err = db.AltSigHistory(ticket.Hash)
+	if err != nil {
+		t.Fatalf("unexpected error for fourth alt history: %v", err)
+	}
+	if !reflect.DeepEqual(wantHist, history) {
+		t.Fatalf("expected history %v but got %v", spew.Sdump(wantHist), spew.Sdump(history))
```

Comment:
Should probably count the history entries here and drop spew as a dependency.

Reply:
aw....

Reply:
ok